### ISSUE TITLE
Does the Binarize() function use STE？

### ISSUE THREAD

**Issue by leejiajun** [itayhubara/BinaryNet.pytorch#9]
Does the Binarize() function use STE？
I haven't seen the STE algorithm in this whole project.

**Comment by haichaoyu** [comment]
Want to know the same +1

**Comment by Ashokvardhan** [comment]
@mohdumar644, @haichaoyu, @leejiajun : I went through the code of `binarized_modules.py` and in the class **BinarizeLinear**, we can see that the line `self.weight.data=Binarize(self.weight.org)` modifies the values of the weights to be binary. This means that both during the forward and backward passes, the binary weights are used. However for the gradient update we need real weights which are only available in `self.weight.org`. However I am not sure how these real weights are used by the optimizer.

**Comment by mohdumar644** [comment]
weights are binarized in forward pass using `self.weight.data=Binarize(self.weight.org)`.

prior to optimizer step, real weights are [copied](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_binary.py#L255) back.

**Comment by mohdumar644** [comment]
btw, a more understandable approach to constructing binarized/quantized neural networks in PyTorch can be found [here ](https://github.com/ussamazahid96/BNN-PYNQ/tree/master/bnn/src/training/Pytorch) with a custom [Quantizer](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/lenet.py#L44)/Binarizer module having [explicit STE](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/binarized_modules.py#L24) in the backward pass.

**Comment by Ashokvardhan** [comment]
@mohdumar644 thanks for the links! In the attached implementation too, the class [BinarizeLinear](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/binarized_modules.py#L53) uses the function [QuantizeWeights](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/binarized_modules.py#L27) which is very similar to this PyTorch implementation. They also use the [copying trick](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/lenet.py#L130). So I don't see a clear difference in the implementation except for the **Quantizer** module which has own forward and backward functions.

**Comment by Ashokvardhan** [comment]
> weights are binarized in forward pass using `self.weight.data=Binarize(self.weight.org)`.
> 
> prior to optimizer step, real weights are [copied](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_binary.py#L255) back.

I see. I was only looking at the file [main_mnist.py](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_mnist.py) which didn't have this copying trick and hence got confused. So it seems the current implementation for MNIST is wrong because they are updating the binary weights instead of the real weights. what do you think?

**Comment by haichaoyu** [comment]
Hi @Ashokvardhan , thanks for your insights. I just focused on the the resnet code for cifar10 (resnet_binary.py). In fact, I have two questions about the code.
- One is the hardtanh [here](https://github.com/itayhubara/BinaryNet.pytorch/blob/master/models/resnet_binary.py#L94). This function is called between bn and next conv layer. hardtanh makes output between -1 and +1. This makes the binarization does not take effects at the beginning of each conv and linear layers. My solution is to just remove the tanh layer and use an explicit STE layer as you mentioned above (The difference is that I used gate function in backward). But after these modifications, performance remains the same.

-   The second possible problem is that the Binarization operation (inside each conv and linear) are directly applied to Tensor.data instead of Tensor. As a result, binarization is not recorded in gradient graph. An explicit STE on Tensor instead of Tensor.data may solve this problem.

Please correct me if there is any problem.

**Comment by mohdumar644** [comment]
> @mohdumar644 thanks for the links! In the attached implementation too, the class [BinarizeLinear](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/binarized_modules.py#L53) uses the function [QuantizeWeights](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/binarized_modules.py#L27) which is very similar to this PyTorch implementation. They also use the [copying trick](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/lenet.py#L130). So I don't see a clear difference in the implementation except for the **Quantizer** module which has own forward and backward functions.

My bad actually since I am a bit rusty - the links I shared only change the way the intermediate activations are binarized/quantized, not the weights - for the weights the way is the same i.e. an extended module.

> > weights are binarized in forward pass using `self.weight.data=Binarize(self.weight.org)`.
> > prior to optimizer step, real weights are [copied](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_binary.py#L255) back.
> 
> I see. I was only looking at the file [main_mnist.py](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_mnist.py) which didn't have this copying trick and hence got confused. So it seems the current implementation for MNIST is wrong because they are updating the binary weights instead of the real weights. what do you think?

Did you look at this [line](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_mnist.py#L115)?

**Comment by mohdumar644** [comment]
> Hi @Ashokvardhan , thanks for your insights. I just focused on the the resnet code for cifar10 (resnet_binary.py). In fact, I have two questions about the code.
> 
> * One is the hardtanh [here](https://github.com/itayhubara/BinaryNet.pytorch/blob/master/models/resnet_binary.py#L94). This function is called between bn and next conv layer. hardtanh makes output between -1 and +1. This makes backward of  binarization inside each conv and linear does not take effects (gradients are always 1). My solution is to just remove the tanh layer and use an explicit STE layer as you mentioned above (The difference is that I used gate function in backward). After these modifications, performance remains the same.

You basically integrated Hardtanh into Quantizer/Binarizer function by using a clamper/gate in the backward. Good!

**Comment by Ashokvardhan** [comment]
> > @mohdumar644 thanks for the links! In the attached implementation too, the class [BinarizeLinear](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/binarized_modules.py#L53) uses the function [QuantizeWeights](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/binarized_modules.py#L27) which is very similar to this PyTorch implementation. They also use the [copying trick](https://github.com/ussamazahid96/BNN-PYNQ/blob/166f2fe40d78468edd0ccaa665d2dcc11aa3dc99/bnn/src/training/Pytorch/lenet.py#L130). So I don't see a clear difference in the implementation except for the **Quantizer** module which has own forward and backward functions.
> 
> My bad actually since I am a bit rusty - the links I shared only change the way the intermediate activations are binarized/quantized, not the weights - for the weights the way is the same i.e. an extended Linear module. The code I shared just avoided having to `input.data = Binarize(input.data)` inside this extended class.
> 
> > > weights are binarized in forward pass using `self.weight.data=Binarize(self.weight.org)`.
> > > prior to optimizer step, real weights are [copied](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_binary.py#L255) back.
> > 
> > 
> > I see. I was only looking at the file [main_mnist.py](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_mnist.py) which didn't have this copying trick and hence got confused. So it seems the current implementation for MNIST is wrong because they are updating the binary weights instead of the real weights. what do you think?
> 
> Did you look at this [line](https://github.com/itayhubara/BinaryNet.pytorch/blob/f5c3672dede608f568e073a583cadd7a8a88fa9d/main_mnist.py#L115)?

Aah, didn't notice this. Thanks! Btw, this file [main_binary.py](https://github.com/itayhubara/BinaryNet.pytorch/blob/master/main_binary.py) assumes that one has access to the datasets for **cifar-10, cifar-100, imagenet**. Do you know in what format they are needed to run this code? In case you already happened to run these codes, can you let me know about this? Thanks!

**Comment by mohdumar644** [comment]
As per [this](https://github.com/itayhubara/BinaryNet.pytorch/blob/master/data.py),

- the cifar-x datasets (and even the mnist dataset when needed) will be automatically downloaded and extracted in a certain folder the first time the script is run, and later reused.
- the imagenet is much larger (>100 gb) and will be needed to be downloaded and setup in the relevant directory yourself

**Comment by Ashokvardhan** [comment]
> Hi @Ashokvardhan , thanks for your insights. I just focused on the the resnet code for cifar10 (resnet_binary.py). In fact, I have two questions about the code.
> 
> * One is the hardtanh [here](https://github.com/itayhubara/BinaryNet.pytorch/blob/master/models/resnet_binary.py#L94). This function is called between bn and next conv layer. hardtanh makes output between -1 and +1. This makes backward of  binarization inside each conv and linear does not take effects (gradients are always 1). My solution is to just remove the tanh layer and use an explicit STE layer as you mentioned above (The difference is that I used gate function in backward). After these modifications, performance remains the same.

Sorry, I didn't understand this. can you please explain in detail?

> * The second possible problem is that the Binarization operation (inside each conv and linear) are directly applied to Tensor.data instead of Tensor. As a result, binarization is not recorded in gradient graph. An explicit STE on Tensor instead of Tensor.data may solve this problem.
> 
> Please correct me if there is any problem.

I realized that there is no problem with their implementation. The reason they didn't have to use STE() is because they are modifying only the Tensor.data which won't be recorded for gradient computation. This is exactly what we need because we are using the gradients on the binary weights to update the real weights directly.

**Comment by haichaoyu** [comment]
Regarding the first point:
- According to the paper (Equation 4), the gradient of STE is 0 when abs(weight) > 0 and 1 when abs(weight) < 1. If activations are passed through torch.HardTanh before STE, abs of all activations are less than 1. Thus gradients of STE are always 1?

Actually, both of my concerns are summarized in this issue title by leejiajun. Does the network use STE with gate gradient? (Equation 4 in the paper)

**Comment by itayhubara** [comment]
By using the HradTanhH  you ensure that the gradients of STE is 0 when abs(activation) > 1 and 1 when abs(activation) < 1. Note that the STE is used over the activations (not the weights) to avoid running back-propagation through the sign function.

**Comment by haichaoyu** [comment]
@itayhubara, Sorry, its my mistake. What about weight STE?

**Comment by Ashokvardhan** [comment]
@mohdumar644  Thanks for the links. Btw, though the paper is about binary weights and activations, in the code I only see weights being binarized. Are activations binarized anywhere? The code uses **nn.Hardtanh()** function which only restricts it to be in [-1,1] but not binary.

**Comment by mohdumar644** [comment]
The line

```
input.data=Binarize(input.data)
```


in binarized_modules.py is related to the activations

**Comment by Ashokvardhan** [comment]
@mohdumar644 : I am confused now because in the backward pass, the gradients for activations are computed at the binary activations and not the real ones. Suppose **h** is the real activation from the previous layer. Ignoring the batch norm stuff, if the real weights are **w**, then the **BinarizeLinear** computes **Hardtanh(Q(w). Q(h))**, where `Q` is the sign quantization function. Now when we do backpass, the gradients of activations are computed at **Q(h)**. Since there is no STE() for activations, I don't see how the current gradients are obtained.

**Comment by mohdumar644** [comment]
> @mohdumar644 : I am confused now because in the backward pass, the gradients for activations are computed at the binary activations and not the real ones. Suppose **h** is the real activation from the previous layer. Ignoring the batch norm stuff, if the real weights are **w**, then the **BinarizeLinear** computes **Hardtanh(Q(w). Q(h))**, where `Q` is the sign quantization function. Now when we do backpass, the gradients of activations are computed at **Q(h)**. Since there is no STE() for activations, I don't see how the current gradients are obtained.

A typical BNN goes like   a_bin_k = Sign(HardTanh(BatchNorm(Linear(w_b * a_bin_k_prev)

The sign binarization function on input.data inside forward function of BinarizeLinear is not recorded in the gradient computation graph.
PyTorch does not see the operation Sign(x), and does not use its gradient. But it has the binarized activations available in the BinarizeLinear saved tensors for backward passed since we did binarize them using sign on input.data. So the gradient on weights g_w_b is calculated using binarized activations, and gradient on binarized activations g_a_b is calculated using binarized weights (according to backprop formulas of linear layer).

Now in our PyTorch model, we defined HardTanh before the next Linear layer, so we backpropagate further from BinarizeLinear to the preceding HardTanh. We did not write Sign function anywhere in our main model description so its gradient is nowhere considered.
If we say y=HardTanh(x), then we already have g_y as g_a_b, since the sign function was not recorded in computation graph. Now we can find g_x and backprop further back. Thus we implicitly performed STE.

**Comment by Ashokvardhan** [comment]
@mohdumar644 Thanks for the clarification. I was under the impression that the gradients with respect to activations are always calculcated w.r.t to real activation values. If they are computed only w.r.t binary activations, indeed the current implementation is current and your explanation makes a lot of sense. With regards to activations being binarized, the other implementation does this explicitly with no source for confusion.

**Comment by mohdumar644** [comment]
> @mohdumar644 Thanks for the clarification. I was under the impression that the gradients with respect to activations are always calculcated w.r.t to real activation values. If they are computed only w.r.t binary activations, indeed the current implementation is current and your explanation makes a lot of sense. With regards to activations being binarized, the other implementation does this explicitly with no source for confusion.

Always good to refer to the original BNN paper.

![image](https://user-images.githubusercontent.com/22178137/60999746-48d60f00-a375-11e9-80ad-50f76736b0d4.png)

**Comment by Ashokvardhan** [comment]
@mohdumar644  I was confused because I didn't see **nn.Hardtanh()** in the paper but in the implementation they used that function. Because of this ambiguity, it was not clear which one to follow. Now looking back, the forward pass in the paper is effectively like a linear neural network just with binary activations and weights and clamping the gradients in the backward pass when activations are beyond 1, right? Of course, for the implementation sake one can view it as using **nn.Hardtanh()** in the forward pass to make the computations of backward pass easier.

**Comment by mohdumar644** [comment]
Yes the paper's algorithm can be functionally achieved using a variety of ways in different frameworks.

**Comment by Ashokvardhan** [comment]
> As per [this](https://github.com/itayhubara/BinaryNet.pytorch/blob/master/data.py),
> 
> * the cifar-x datasets (and even the mnist dataset when needed) will be automatically downloaded and extracted in a certain folder the first time the script is run, and later reused.
> * the imagenet is much larger (>100 gb) and will be needed to be downloaded and setup in the relevant directory yourself

I see. Do you need to do any processing on the imagenet dataset or just the raw images are used for the code? Btw, where did you downloaded it from?

**Comment by Ashokvardhan** [comment]
@itayhubara Can you mention a bit what the instructions for ImageNet dataset that are needed for the experiments in the paper? Such as what format you need the dataset and where to obtain it from, etc.

**Comment by codeprb** [comment]
@itayhubara I too have the same question as @Ashokvardhan. Which imagenet dataset has been used here. I have downloaded a partial dataset from kaggle with 45000 images for training and 5000 images for validation. Please let me know if there is any specific requirement for imagenet dataset.

**Comment by itayhubara** [comment]
You need to download the ImageNet dataset from here http://www.image-net.org/challenges/LSVRC/2012/downloads (note that you must login for that) and then you can use torchvision's to create a dataloader.

**Comment by codeprb** [comment]
@itayhubara thanks for the reply. I am implementing this code for resnet_binary. could you please let me know if it uses binary weights as well as activation for renet_binary code.

**Comment by codeprb** [comment]
I am trying to run Resnet model with the Imagenet dataset. Unfortunately, I get 100% accuracy soon after the 1 epoch of the training. I am unable to trace the error. Could you please let me know what could be the error.

I am storing data under datasets/ImageNet_train/train and test data as datasets/ImageNet_test/test

**Comment by yananliusdu** [comment]
> @mohdumar644 Thanks for the links. Btw, though the paper is about binary weights and activations, in the code I only see weights being binarized. Are activations binarized anywhere? The code uses **nn.Hardtanh()** function which only restricts it to be in [-1,1] but not binary.
> 
> In fact, the other [implementation](https://github.com/ussamazahid96/BNN-PYNQ/tree/master/bnn/src/training/Pytorch) you referred to actually uses quantization on activation.

that's true. also I noticed in another version of BNN, they implement binarization after activations. [https://github.com/Akashmathwani/Binarized-Neural-networks-using-pytorch/blob/master/MNIST%20using%20Binarized%20weights/modules.py](url)
